### PR TITLE
DataGridNext - virtual scrolling

### PR DESCRIPTION
**Pull request by NickMitrokhin**

### CODE REVIEW

File: js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    const skip = this.plugins.getValue(ViewportSkipValue) ?? 0;
+    const take = this.plugins.getValue(ViewportTakeValue) ?? 0;
+    const pageIndex = Math.floor(pageSize > 0 ? skip / pageSize : 0);
+    const pageOffset = pageIndex * pageSize;
+    const skipForCurrentPage = skip - pageOffset;
+    const loadPageCount = Math.ceil(pageSize > 0 ? (take + skipForCurrentPage) / pageSize : 0);
+
+    this.plugins.callAction(SetPageIndex, pageIndex);
+    this.plugins.callAction(SetLoadPageCount, loadPageCount);
+  }
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx]:
```suggestion
  loadViewport(): void {
    const pageSize = this.plugins.getValue(PageSize) ?? 'all';
    if (pageSize !== 'all') {
      const skip = this.plugins.getValue(ViewportSkipValue) ?? 0;
      const take = this.plugins.getValue(ViewportTakeValue) ?? 0;
      const pageIndex = Math.floor(skip / pageSize);
      const pageOffset = pageIndex * pageSize;
      const skipForCurrentPage = skip - pageOffset;
      const loadPageCount = Math.ceil((take + skipForCurrentPage) / pageSize);

      this.plugins.callAction(SetPageIndex, pageIndex);
      this.plugins.callAction(SetLoadPageCount, loadPageCount);
    }
  }
```

File: js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    if (this.scrollPosition.top !== top || this.scrollPosition.left !== left) {
+      this.scrollPosition = { top, left };
+      this.loadViewport();
+    }
+  }
+
+  loadViewport(): void {
+    let pageSize = this.plugins.getValue(PageSize) ?? 0;
+    pageSize = typeof pageSize !== 'number' ? 0 : pageSize;
+    const skip = this.plugins.getValue(ViewportSkipValue) ?? 0;
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx]: Try to avoid using of `.getValue`, use `.watch` instead

**Review by pomahtri — CHANGES_REQUESTED** [review]

File: js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx (hunk 1 above)
Comment (NickMitrokhin) [reply js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx]: I think it is better to recalculate pageIndex and loadPageCount each time the loadViewport is called. Especially if pageSize is changed from any number to "all".

File: js/renovation/ui/grids/data_grid_next/views/table_content.tsx [hunk 3]
```diff
@@ -71,11 +102,22 @@ export class TableContent extends JSXComponent(TableContentProps) {
     return (): void => eventsEngine.off(this.divRef.current, clickEvent, onRowClick);
   }
 
+  @Effect()
+  calculateRowsViewHeight(): void {
+    this.plugins.set(RowsViewHeightValue, getElementHeight(this.rowsViewRef.current));
+  }
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_next/views/table_content.tsx:116]: Maybe use resize observer cause height won't be updated on its change

File: js/renovation/ui/grids/data_grid_next/scrolling/scrolling.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+export const viewFunction = ({
+  isVirtualScrolling, virtualScrollingMode,
+  props: {
+    mode,
+  },
+}: DataGridNextScrolling): JSX.Element => (
+  <Fragment>
+    <ValueSetter type={ScrollingModeValue} value={mode} />
+    {isVirtualScrolling && (<VirtualScrolling mode={virtualScrollingMode} />)}
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_next/scrolling/scrolling.tsx:18]:
```suggestion
    {mode !== 'standard' && (<VirtualScrolling mode={mode} />)}
```

In this case there is no need in confusing getter `virtualScrollingMode`

File: js/renovation/ui/grids/data_grid_next/scrolling/virtual_content.tsx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    />
+    { topHeight > 0 && (
+      <PlaceholderExtender
+        type={TopRowPlaceholder}
+        order={1}
+        template={(): JSX.Element => (
+          <VirtualRow
+            height={topHeight}
+            cellClasses={cellClasses}
+            rowKey={-1}
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_next/scrolling/virtual_content.tsx:42]: no need in react key here, cause Virtual Row is not in elements array

File: js/renovation/ui/grids/data_grid_next/scrolling/virtual_content.tsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  @InternalState()
+  bottomHeight = 0;
+
+  @InternalState()
+  cellClasses: string[] = [];
+
+  @Effect()
+  watchVisibleColumns(): () => void {
+    return this.plugins.watch(VisibleColumns, (visibleColumns) => {
+      this.cellClasses = Array(visibleColumns.length).fill('').map((_, index) => visibleColumns[index].headerCssClass ?? '');
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_next/scrolling/virtual_content.tsx:87]: Why we need to use classes from headers?

File: js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+  @InternalState()
+  viewportState: ViewportStateType = 'synchronized';
+
+  @Mutable()
+  viewportPayload = { pageIndex: 0, topScrollPosition: 0 };
+
+  get itemHeights(): Record<number, number> {
+    return this.visibleItemHeights;
+  }
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx:95]: Why using getter that just returns mutable?

**Review by pomahtri — DISMISSED** [review]
I left some notes and questions as checklist for future, I think it's better to handle them later in next PR

Bugs:

- After fast scrolling from bottom to the very top rows don't render

File: js/renovation/ui/grids/data_grid_next/plugins.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  ],
+  (
+    dataStateValue: DataState, keyExpr: KeyExprInternal,
+  ): Row[] => {
+    const rows = dataStateValue.data.map((data) => ({
+      key: keyExpr ? data[keyExpr] : data,
+      data,
+      rowType: 'data',
+    }));
+    return rows as Row[];
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_next/plugins.ts:40]:
```suggestion
    const rows = dataStateValue.data.map((data): Row => ({
      key: keyExpr ? data[keyExpr] : data,
      data,
      rowType: 'data',
    }));
    return rows;
```

Comment (pomahtri) [reply js/renovation/ui/grids/data_grid_next/plugins.ts:40]: Also need to figure out how to get rid of other `as`

File: js/renovation/ui/grids/data_grid_next/scrolling/virtual_content.tsx (hunk 6 above)
Comment (NickMitrokhin) [reply js/renovation/ui/grids/data_grid_next/scrolling/virtual_content.tsx:87]: In order to align virtual cells according to columns

File: js/renovation/ui/grids/data_grid_next/scrolling/virtual_content.tsx (hunk 5 above)
Comment (NickMitrokhin) [reply js/renovation/ui/grids/data_grid_next/scrolling/virtual_content.tsx:42]: This key is required to render virtual rows among data rows

File: js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx (hunk 7 above)
Comment (NickMitrokhin) [reply js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx:95]: In order to share this value among other Selectors in virtual scrolling plugins.

**Review by Alyar666 — DISMISSED** [review]

**Pull request merged**